docs(pt-br): translated passing-props-to-a-component.md (#676)
reactjs/pt-br.react.dev: src/content/learn/passing-props-to-a-component.md

</Sandpack>

As props que você pode passar a uma tag `<img>` são predefinidas (A ReactDOM conforma-se ao [padrão HTML](https://www.w3.org/TR/html52/semantics-embedded-content.html#the-img-element)). Mas você pode passar quaisquer props aos *seus próprios* componentes, como um `<Avatar>`, para customizá-los. Veja como fazer isso!

## Passando props para um componente {/*passing-props-to-a-component*/}

> Neste código, o componente `Profile` não está passando nenhuma prop aos seus componente filhos, `Avatar`:

Reviewer: aqui está se referindo à apenas um componente filho - Avatar, então ao invés do plural `aos seus componente filhos` melhor seria no singular `ao seu componente filho`.

```suggestion
Neste código, o componente `Profile` não está passando nenhuma prop ao seu componente filho, `Avatar`:
```

}
```

Você pode atribuir algumas props ao `Avatar` em dois passos.

> ### Passo 1: Passe props aos componentes filhos {/*step-1-pass-props-to-the-child-component*/}

Reviewer: aqui também seria no singular

```suggestion
### Passo 1: Passe props ao componente filho {/*step-1-pass-props-to-the-child-component*/}
```

<Note>

> Se as chaves duplas depois de `person=` lhe confundem, lembre-se [que elas são meramente um objeto](/learn/javascript-in-jsx-with-curly-braces#using-double-curlies-css-and-other-objects-in-jsx) na sintaxe JSX.

Reviewer: ```suggestion
Se as chaves duplas depois de `person=` confundir você, lembre-se [que elas são meramente um objeto](/learn/javascript-in-jsx-with-curly-braces#using-double-curlies-css-and-other-objects-in-jsx) dentro das chaves do JSX.
```

Author: Uma coisa que eu sinto bastante na tradução dessas páginas é a repetição frequente de "você", por isso em alguns contextos eu tento dar uma enxugada nessa repetição visto que a nossa língua tem mais recursos gramaticais para fazer isso.
Já na parte "dentro das chaves do JSX", creio que se perca o sentido original da frase.

Reviewer: Acredito que devemos utilizar uma linguagem mais simples e direta pensando em quem está lendo, ou seja, menos formal. E justamente se em todo lugar está utilizando `você` para se dirigir ao leitor, melhor seria continuar nessa mesma linha.

Sobre as chaves, não creio que perca o sentido porque justamente está especificando exatamente onde está o objeto - é um objeto dentro das chaves de JSX (talvez poderia trocar `do JSX` por `de JSX` aqui). Se você coloca "na sintaxe JSX", perde esta especificação - fica muito geral e não especifica onde está o objeto - perde o sentido da explicação do porquê das chaves duplas.

Author: Você tem um ponto, só vou alterar a conjugação do verbo "confundir" ali
```suggestion
Se as chaves duplas depois de `person=` confundirem você, lembre-se [que elas são meramente um objeto](/learn/javascript-in-jsx-with-curly-braces#using-double-curlies-css-and-other-objects-in-jsx) dentro das chaves da JSX.
```

</Note>

Agora você pode ler essas props dentro do componente `Avatar`.

### Passo 2: Leia props dentro de um componente filho {/*step-2-read-props-inside-the-child-component*/}

> Você pode ler estas propriedades listando seus nomes `person, size` separados por vírgulas diretamente dentro de `({` e `})` depois de `function Avatar`. Isso permite que você as use dentro do código de `Avatar`, assim como você faria com uma variável.

Reviewer: ```suggestion
Você pode ler estas props listando seus nomes `person, size` separados por vírgulas dentro de `({` e `})` diretamente depois de `function Avatar`. Isso permite que você as use dentro do código de `Avatar`, assim como você faria com uma variável.
```

</Sandpack>

> Props permitem que você pense sobre os componentes pai e filho independentemente. Por exemplo, você pode mudar as props `person` e `size` dentro de `Profile` sem ter que pensar sobre como `Avatar` as usa. Similarmente, você é livre para mudar como `Avatar` usa essas props, sem checar o `Profile`.

Reviewer: ```suggestion
Props permitem que você pense sobre os componentes pai e filho independentemente. Por exemplo, você pode mudar as props `person` ou `size` dentro de `Profile` sem ter que pensar sobre como `Avatar` as usa. Similarmente, você pode mudar como `Avatar` usa essas props, sem olhar para o `Profile`.
```

Author: A troca do "e" pelo "ou" realmente ficou melhor!
Quanto ao restante, a tradução literal realmente seria essa porém tomei a liberdade de alterar um pouco a frase original para não repetir tanto os termo "pode/poder".
```suggestion
Props permitem que você pense sobre os componentes pai e filho independentemente. Por exemplo, você pode mudar as props `person` ou `size` dentro de `Profile` sem ter que pensar sobre como `Avatar` as usa. Similarmente, você é livre para mudar como `Avatar` usa essas props, sem checar o `Profile`.
```

> Você pode pensar nas props como "controles" os quais você pode ajustar. Elas desempenham o mesmo papel que os argumentos para funções--de fato, props _são_ o único argumento para o seu componente! Os componente funcionais do React aceitam apenas um argumento, um objeto `props`:

Reviewer: ```suggestion
Você pode pensar nas props como "controles" os quais você pode ajustar. Elas desempenham o mesmo papel que os argumentos para funções-de fato, props _são_ o único argumento para o seu componente! Os componente funcionais do React aceitam apenas um argumento, um objeto `props`:
```

}
```

Normalmente você não precisa de todo o objeto `props` em si, então você pode desestruturá-lo em props individuais.

<Pitfall>

> **Não esqueça o par `{` e `}`** dentro de `(` e `)` ao declarar props:

Reviewer: ```suggestion
**Não esqueça o par de `{` e `}` chaves** dentro de `(` e `)` ao declarar props:
```

```js
function Avatar({ person, size }) {
  // ...
}
```

> Essa sintaxe é chamada de ["desestruturação"](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment#Unpacking_fields_from_objects_passed_as_a_function_parameter) e é equivalente a ler propriedades de um parâmetro de função:

Reviewer: ```suggestion
Esta sintaxe é chamada de ["desestruturação"](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment#Unpacking_fields_from_objects_passed_as_a_function_parameter) e é equivalente a ler propriedades de um parâmetro de função:
```

</Pitfall>

> ## Especificando um valor padrão a uma prop {/*specifying-a-default-value-for-a-prop*/}

Reviewer: ```suggestion
## Especificando um valor padrão para uma prop {/*specifying-a-default-value-for-a-prop*/}
```

> Se você quer dar a uma prop um valor padrão para usar quando nenhum for especificado, pode fazer isso com a desestruturação colocando `=` e um valor padrão logo depois do parâmetro:

Reviewer: ```suggestion
Se você quer dar à prop um valor padrão para usar quando nenhum valor for especificado, você pode fazer isso com a desestruturação colocando `=` e o valor padrão logo depois do parâmetro:
```

Author: O uso da crase implica que estamos nos referindo a uma prop específica, o que não é o caso neste contexto. Também discordo com a repetição do "você", mais uma vez usando nossa gramática rica a nosso favor.
```suggestion
Se você quer dar a uma prop um valor padrão para usar quando nenhum valor for especificado, pode fazer isso com a desestruturação colocando `=` e o valor padrão logo depois do parâmetro:
```

```js
function Avatar({ person, size = 100 }) {
  // ...
}
```

> Agora, se `<Avatar person={...} />` for renderizado sem uma prop `size`, `size` será igual a `100`.

Reviewer: ```suggestion
Agora, se `<Avatar person={...} />` for renderizado sem a prop `size`, `size` será igual a `100`.
```

> O valor padrão só é utilizado se a prop `size` não for especificada ou se você passar `size={undefined}`. Mas caso você passe `size={null}` ou `size={0}`, o valor padrão **não** será usado.

Reviewer: ```suggestion
O valor padrão é apenas utilizado se a prop `size` não for especificada ou se você passar `size={undefined}`. Mas caso você passe `size={null}` ou `size={0}`, o valor padrão **não** será usado.
```

}
```

> Não há nada de errado com o código repetitivo--ele pode ser mais legível. Mas por vezes você pode valorizar concisão. Alguns componentes encaminham todas as suas props aos seus filhos, como `Profile` faz com `Avatar`. Como ele não usa nenhuma de suas props diretamente, pode fazer sentido o uso da mais sucinta sintaxe de espalhamento:

Reviewer: ```suggestion
Não há nada de errado com código repetitivo-ele pode ser mais legível. Mas às vezes você pode valorizar concisão. Alguns componentes encaminham todas as suas props aos seus filhos, como `Profile` faz com `Avatar`. Como eles não usam nenhuma de suas props diretamente, pode fazer sentido usar uma sintaxe de espalhamento mais concisa:
```

}
```

Isso encaminha todas as props de `Profile` ao `Avatar` sem listar cada um de seus nomes.

> **Use a sintaxe de espalhamento com cuidado.** Se você está a utilizando em quase todos os componentes, algo está errado. Geralmente, isso indica que você deveria separar seus componentes e passar filhos como JSX. Falaremos mais sobre isso agora!

Reviewer: ```suggestion
**Use a sintaxe de espalhamento com cuidado.** Se você está a utilizando em quase todos os componentes, algo está errado. Muitas vezes, isso indica que você deveria dividir seus componentes e passar filhos como JSX. Mais sobre isso a seguir!
```

> ## Passando JSX como filhos {/*passing-jsx-as-children*/}

Reviewer: Aqui estou pensando se é melhor deixar o termo `children` em inglês mesmo pois abaixo na explicação se utiliza muito `children` ao invés de filho. O que você acha?

```suggestion
## Passando JSX como children {/*passing-jsx-as-children*/}
```

Author: É uma boa ideia, porém acredito que fique ainda melhor caso coloquemos children dentro de um bloco:
```suggestion
## Passando JSX como `children` {/*passing-jsx-as-children*/}
```

É comum aninhar tags embutidas no navegador:

```js
<div>
  <img />
</div>
```

> Às vezes será interessante aninhar seus componentes da mesma forma:

Reviewer: ```suggestion
Às vezes será interessante aninhar seus próprios componentes da mesma forma:
```

Author: ```suggestion
Às vezes você desejará aninhar seus próprios componentes da mesma forma:
```

```js
<Card>
  <Avatar />
</Card>
```

> Quando você aninha conteúdo dentro de uma tag JSX, o componente pai irá recebê-lo dentro de uma prop denominada `children`. Exemplificando, o componente `Card` abaixo receberá a prop `children` definida como `<Avatar />` e o renderizará em uma div a qual o encapsula:

Reviewer: ```suggestion
Quando você aninha conteúdo dentro de uma tag JSX, o componente pai irá recebê esse conteúdo em uma prop chamada `children`. Por exemplo, o componente `Card` abaixo receberá a prop `children` definida como `<Avatar />` e o renderizará em uma div a qual o encapsula:
```

> Quando você aninha conteúdo dentro de uma tag JSX, o componente pai irá recebê esse conteúdo em uma prop chamada `children`. Por exemplo, o componente `Card` abaixo receberá a prop `children` definida como `<Avatar />` e o renderizará em uma div a qual o encapsula:

Reviewer: Olá @TarsiSurdi, apenas mais um pequeno detalhe aqui. De acordo com o [glossário](https://github.com/reactjs/pt-br.react.dev/blob/main/GLOSSARY.md) não deve-se traduzir a palavra `wrapper`.

```suggestion
Quando você aninha conteúdo dentro de uma tag JSX, o componente pai irá receber esse conteúdo em uma prop chamada `children`. Por exemplo, o componente `Card` abaixo receberá a prop `children` definida como `<Avatar />` e o renderizará em uma wrapper div:
```

Author: Obrigado! Me confundi entre "to wrap" (que deve ser traduzido) e "wrapper" :(

</Sandpack>

Tente substituir o `<Avatar>` dentro de `<Card>` com algum texto para ver como o componente `Card` pode encapsular conteúdo aninhado. Ele não precisa "saber" o que está sendo renderizado dentro dele. Você encontrará esse padrão flexível em muitos lugares.

> É possível pensar sobre um componente com a prop `children` como se ele tivesse um "buraco" o qual pode ser "preenchido" por seus componente pais com JSX arbitrária. Você frequentemente usará a prop `children` para encapsuladores visuais: painéis, grids, etc.

Reviewer: ```suggestion
É possível pensar sobre um componente com a prop `children` como se ele tivesse um "buraco" o qual pode ser "preenchido" por seus componente pais com JSX arbitrária. Você frequentemente usará a prop `children` para wrappers visuais: painéis, grids, etc.
```

<Illustration src="/images/docs/illustrations/i_children-prop.png" alt='A puzzle-like Card tile with a slot for "children" pieces like text and Avatar' />

## Como props mudam com o passar do tempo {/*how-props-change-over-time*/}

> O componente `Clock` abaixo recebe duas props de seu componente pai: `color` e `time`. (O código deste componente pai está omitido porque usa [state](/learn/state-a-components-memory), conceito o qual nós ainda não abordamos.)

Reviewer: ```suggestion
O componente `Clock` abaixo recebe duas props de seu componente pai: `color` e `time`. (O código deste componente pai está omitido porque usa [state](/learn/state-a-components-memory), conceito o qual nós não vamos nos aprofundar ainda.)
```

</Sandpack>

Este exemplo ilustra que **um componente pode receber props diferentes com o passar o tempo.** Props não são sempre estáticas! Aqui, a prop `time` muda a cada segundo, e a prop `color` muda quando você seleciona outra cor. As props refletem os dados de um componente a qualquer instante, não apenas num primeiro momento.

> Entretanto, as props são [imutáveis](https://pt.wikipedia.org/wiki/Objeto_imut%C3%A1vel)--um termo da ciência da computação o qual significa "inalterável". Quando um componente precisa mudar suas props (por exemplo, em resposta à interação do usuário ou a novos dados), ele terá que "pedir" ao componente pai que passe _props diferentes_-- um novo objeto! Essas props antigas serão então deixadas de lado, e eventualmente o motor do JavaScript irá recuperar a memória ocupada por elas.

Reviewer: ```suggestion
Entretanto, as props são [imutáveis](https://pt.wikipedia.org/wiki/Objeto_imut%C3%A1vel)-um termo da ciência da computação o qual significa "inalterável". Quando um componente precisa mudar suas props (por exemplo, em resposta à interação do usuário ou a novos dados), ele terá que "pedir" ao componente pai que passe _props diferentes_- um novo objeto! Suas props antigas serão então deixadas de lado, e eventualmente o motor do JavaScript irá recuperar a memória ocupada por elas.
```

> **Não tente "alterar props".** Quando você precisa responder a interações do usuário (como trocar a cor selecionada), você terá que "definir state", sobre o qual você pode aprender sobre em [State: A Memória de um Componente.](/learn/state-a-components-memory)

Author: ```suggestion
**Não tente "alterar props".** Quando você precisa responder a interações do usuário (como trocar a cor selecionada), você terá que "definir state", sobre o qual você pode aprender em [State: A Memória de um Componente.](/learn/state-a-components-memory)
```

<Recap>

* Para passar props, adicione-as à JSX, assim como você faria com atributos HTML.
* Para ler props, use a sintaxe de desestruturação `function Avatar({ person, size })`.
* Você pode especificar um valor padrão como `size = 100`, o qual é usado para props inexistentes ou `undefined`.

> * Você pode encaminhar todas as props com a sintaxe de espalhamento JSX `<Avatar {...props} />`, mas não a utilize exaustivamente.

Reviewer: ```suggestion
* Você pode encaminhar todas as props com a sintaxe de espalhamento JSX `<Avatar {...props} />`, mas não abuse!
```

* JSX aninhada como `<Card><Avatar /></Card>` aparecerá como a prop `children` do componente `Card`.
* Props podem somente ser lidas e representam um momento específico no tempo: toda renderização recebe uma nova versão de props.

> * Props são imutáveis. Quando você precisar de interatividade, defina state.

Reviewer: ```suggestion
* Você não pode mudar as props. Quando você precisar de interatividade, precisará definir o state.
```

Author: ```suggestion
* Você não pode mudar as props. Quando você precisar de interatividade, precisará definir state.
```

</Recap>

<Challenges>

#### Extraia um componente {/*extract-a-component*/}

> Este componente `Gallery` contém marcação bastante similar para os dois perfis. Extraia um componente `Profile` a partir dele para reduzir a duplicação de código. Você precisará escolher quais props o passar.

Reviewer: ```suggestion
Este componente `Gallery` contém marcação bastante similar para os dois perfis. Extraia um componente `Profile` a partir dele para reduzir a duplicação de código. Você precisará escolher quais props passar para ele.
```

</Sandpack>

Mesmo que a sintaxe seja levemente diferente porque você está descrevendo as propriedades de um objeto JavaScript em vez de uma coleção de atributos JSX, esses exemplos são praticamente equivalentes e você pode escolher ambas as abordagens.

</Solution>

> #### Ajuste o tamanho da imagem com base um uma prop {/*adjust-the-image-size-based-on-a-prop*/}

Reviewer: ```suggestion
#### Ajuste o tamanho da imagem com base em uma prop {/*adjust-the-image-size-based-on-a-prop*/}
```

> Neste exemplo, `Avatar` recebe uma prop numérica `size` a qual determina a largura e a altura da `<img>`. A prop `size` é igual a `40` neste exemplo. No entanto, se você abri-la em uma nova aba, perceberá que a imagem em si é maior (`160` pixels). O tamanho real da imagem é determinado por qual miniatura você está requisitando.

Reviewer: ```suggestion
Neste exemplo, `Avatar` recebe uma prop numérica `size` a qual determina a largura e a altura da `<img>`. A prop `size` é igual a `40` neste exemplo. No entanto, se você abrir a imagem em uma nova aba, perceberá que a ela em si é maior (`160` pixels). O tamanho real da imagem é determinado pelo tamanho da miniatura que você está solicitando.
```

> Altere o componente `Avatar` para requisitar o tamanho mais próximo da imagem com base na prop `size`. Mais especificamente, se `size` for menor que `90`, passe `'s'` ("small") em vez de `'b'` ("big") à função `getImageUrl`. Verifique se suas mudanças funcionam renderizando avatares com diferentes valores da prop `size` e abrindo as imagens em uma nova aba.

Reviewer: ```suggestion
Altere o componente `Avatar` para solicitar o tamanho mais próximo da imagem com base na prop `size`. Especificamente, se `size` for menor que `90`, passe `'s'` ("small") em vez de `'b'` ("big") à função `getImageUrl`. Verifique se suas mudanças funcionam renderizando avatares com diferentes valores da prop `size` e abrindo as imagens em uma nova aba.
```

</Sandpack>

> Props permitem o encapsulamento de lógica, como a interna ao componente `Avatar` (e altere-a depois caso necessário), para que todos possam usar `<Avatar>` sem pensar sobre como as imagens são requisitadas e redimensionadas.

Reviewer: ```suggestion
Props permitem encapsular a lógica como esta dentro do componente `Avatar`, (e alterá-la depois caso necessário), para que todos possam usar o componente `<Avatar>` sem pensar sobre como as imagens são solicitadas e redimensionadas.
```

<Hint>

> Qualquer JSX que você inserir dentro da tag do componente será passada como a prop `children` desse.

Reviewer: ```suggestion
Qualquer JSX que você colocar dentro da tag do componente será passada como prop `children` para o mesmo.
```

</Hint>

<Solution>

> Esta é a maneira com que você pode usar o componente `Card` em ambos os lugares:

Reviewer: ```suggestion
Esta é a maneira que você pode usar o componente `Card` em ambos os lugares:
```
